todo_app/app.py

```diff
@@ -25,25 +30,73 @@ def create_app():
             format="%(asctime)s %(levelname)s %(name)s %(threadName)s: %(message)s",
         )
 
+    app_user = User()
+
+    # OAuth Login
+    login_manager = LoginManager()
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        # logic to redirect to GH auth flow if unauthed
+
+        client_id: str = os.environ.get("GH_CLIENTID")
+        callback_uri = os.environ.get("CALLBACK_URI")
+
+        return redirect(
+            f"https://github.com/login/oauth/authorize?client_id={client_id}&redirect_uri={callback_uri}"
+        )
+
+    @login_manager.user_loader
+    def load_user(user_id: str) -> User:
+        app_user = User()
+        return app_user
+
+    login_manager.init_app(app)
+
+    # MongoDB connection and setup
     mongodbConnectionString = os.environ.get("MONGO_CONN_STRING")
     applicationDatabase = os.environ.get("MONGO_DB_NAME")
 
-    todo = AppDatabase(mongodbConnectionString).connectDatabase(applicationDatabase)
+    # todo = AppDatabase(mongodbConnectionString).connectDatabase(applicationDatabase)
+    app_db = AppDatabase(
+        mongodbConnectionString,
+        database_name=applicationDatabase,
+        collection_name="todo",
+    )
+    user_db = AppDatabase(
+        mongodbConnectionString,
+        database_name=applicationDatabase,
+        collection_name="auth_users",
+    )
 
     @app.route("/")
+    @login_required
     def index():
 
-        # MongoClient document finding
-        collection = todo.todo
-        documents = collection.find()
-        #
+        app_db = AppDatabase(
+            mongodbConnectionString,
+            database_name=applicationDatabase,
+            collection_name="todo",
+        )
+        card_list: list = app_db.get_items()
+        card_list_view_model = ViewModel(card_list)
+        card_list_view_model.user_role = app_user.role
+
+        return render_template("index.html", view_model=card_list_view_model)
 
-        cardList: list = getItems(documents)
-        cardList_view_model = ViewModel(cardList)
+    def writer_required(func):
```

Comment:
This isn't working as we currently stand - try allowing your user to see the button on the index, and you'll see the abort message doesn't appear! The immediate reason appears to be because (highly annoyingly!) Flask insists that the `@app.route(...)` decorator is always the first one, so currently the `writer_required` decorator doesn't seem to be functioning - try switching the order of those two on the `/add` route and see the difference when you try to bypass your system and add a to-do